File: AIS/ais.m
clear;
clc;
%***************** ОТКРЫТИЕ ФАЙЛА *****************************************
fid = fopen('ais138.dat', 'rb');  % открытие файла на чтение
if fid == -1
    error('File is not opened');
end

%f=fopen('my.txt','wt');

frame=0;             % инициализация переменной
cnt=1;              % инициализация счетчика
while ~feof(fid)    % цикл, пока не достигнут конец файла
    [V,N] = fread(fid, 1, 'float');  %считывание одного
% значения float (V содержит значение
% элемента, N – число считанных элементов)
    if N > 0        % если элемент был прочитан успешно, то
        frame(cnt)=V;% формируем вектор-строку из значений V     
        %fprintf(f, '%f', V);
        %fprintf(f, ' ');
        cnt=cnt+1;  % увеличиваем счетчик на 1
    end
end

%fclose(f);
% disp(data);        % отображение результата на экран
% plot(frame);
% length(frame)
fclose(fid);        % закрытие файла 

%***************** ПРЕОБРАЗОВАНИЕК ДИФФ. ВИДУ *****************************
%Передискретизация в 4 раза
re1 = frame(1: 8: end);
im1 = frame(2: 8: end);
re2 = frame(3: 8: end);
im2 = frame(4: 8: end);
re3 = frame(5: 8: end);
im3 = frame(6: 8: end);
re4 = frame(7: 8: end);
im4 = frame(8: 8: end);
signal1 = complex(re1, im1);
signal2 = complex(re2, im2);
signal3 = complex(re3, im3);
signal4 = complex(re4, im4);

length_signal = 4*length(signal1);
signal = zeros(1, length_signal);

k = 1 : 4 : length(signal);
i = 1 : length(signal1);

signal(k) = signal1(i);
signal(k+1) = signal2(i);
signal(k+2) = signal3(i);
signal(k+3) = signal4(i);

% scatterplot(signal)
l_s = length(signal);
diff_signal = zeros(1,l_s);

%***1ый ВАРИАНТ***%
row = 1 : length(signal) - 1;
% diff_signal(row) = signal(row) .* conj(signal(1));

%***2ой ВАРИАНТ***%
diff_signal(row) = signal(row) .* conj(signal(row+1));

% l_diff = length(diff_signal)
% scatterplot(diff_signal)

%***************** ПОДГОТОВКА ПРЕАМБУЛЫ ***********************************
preamble = [0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 1 1 1 1 1 0];
%preamble = [1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 0 1 1 1 1 1 1 0];
length_data = length(signal) - 1;
length_preamble = length(preamble);
  
twiddles = zeros(1, 17);
k = 1 : 17;
twiddles(k) = exp(1i * (k - 1) * pi/8);

synchro = [twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5),twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:17), twiddles(2:9), twiddles(8:-1:5)];
%synchro = [twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5),twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(2:5), twiddles(4:-1:1), twiddles(17:-1:14), twiddles(15:16), twiddles(1:2), twiddles(3:14), twiddles(15:17), twiddles(2:6), twiddles(5:-1:2)];

l_syn = length(synchro);

%**************************** БЫСТРАЯ КОРРЕЛЯЦИЯ **************************
% Длина фрейма
length_frame = 1024;
% Длина пакета данных
length_data = 1024;
% Вспомгательная переменная
i = 1;
% Заполняем до конца преамбулу нулями
synchro1 = [synchro, zeros(1,length_frame-length(synchro))];

p = 1 : length_frame : l_s; 
kol_frame = length(p);

for k = 1 : length_frame : l_s;    
    k_corr(i,:) = ifft(fft(signal(k:(k+length_frame-1))).*conj(fft(synchro1)));
    [max_corr(i),id(i)] = max(k_corr(i,:));
    i = i +1;
end
k_corr2 = k_corr(:);
% plot(abs(k_corr2))
i = 1 : kol_frame;
norm_max_corr(i) = max_corr(i)./max(max(max_corr(i)));

[m,i] = max(norm_max_corr);
start_package = (i-1)*length_frame + id(i);
% Выделение пакета данных
data = signal(start_package : (start_package + length_data - 1));
%row = 1 : 1023;
%data(row) = data(row) ./ conj(data(row+1));
%************************** ДЕМОДУЛЯТОР ***********************************
sig1= [1 exp(1i*pi/8) exp(1i*pi/4) exp(1i*3*pi/8) 1i]; %[ 1 exp(1i*pi/4) 1i exp(1i*3*pi/4) 1]; 
sig2=conj(sig1);
k=1;

bits = uint8(zeros(1, length(data')/4));
for index=3:4:length(data)-5
x=[data(index),data(index+1),data(index+2),data(index+3),data(index+4)];
%Products-- solver outputs
pr1=0;
pr2=0;

%const=conj(data(index))/abs(data(index));
for step=1:5
    pr1=pr1+x(step)*sig1(step);
    pr2=pr2+x(step)*sig2(step);
end;
if real(pr1) > real(pr2)
        bits(k) = 0;
    else
        bits(k) = 1;
end;
k=k+1;
end;

% Для первых 32 бит
for y = 1 : 32
if bits(y) == bits(y+1)
    bits(y) = 1;
else 
    bits(y) = 0;
end;
end;

%***************************** ПАРСИНГ ************************************
coordinate_data = bits(33:200);
l_c = length(coordinate_data);

FCR_data = bits(201:216);
l_FCR = length(FCR_data);

%************************ ПОДСЧЁТ CRC *************************************
summ = bitxor(coordinate_data(153:168), FCR_data);

order = 0;
polynomial = 0;
m = l_FCR;
syms x;
for t = l_c : -1 : (l_c-l_FCR+1)
    polynomial = polynomial + (x^order)*summ(m);
    order = order + 1;
    m = m - 1;
end
for k = (l_c-l_FCR) : -1 : 1
    polynomial = polynomial + (x^order)*coordinate_data(k);
    order = order + 1;
end;

% order_FCR = 0;
% poly_FCR = 0;
% syms x;
% for e = l_FCR : -1 : 1
%     poly_FCR = poly_FCR + (x^order_FCR)*FCR_data(e);
%     order_FCR = order_FCR + 1;
% end;

CRC = 0;
order_CRC= 0;
for h = 16 : -1 : 1
    CRC = CRC + x^order_CRC;
    order_CRC = order_CRC + 1;
end;

vector_data = [1];
CRC_data = [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1];
coordinate_data(153:168) = summ;
for q = 1 : 168
    vector_data(q) = coordinate_data(q);
end

[q,r] = deconv(vector_data, CRC_data);

%************************** ПЕРЕВОД ИЗ 2ой в 10ую *************************
%*** ID
id_data = coordinate_data(1:32);
ID = 0;
for u = 32 : -1 : 1
    ID = ID + 2^(id_data(u));
end;
ID

%*** latitude
latitude_data = coordinate_data(33:64);
latitude = 0;
for u = 32 : -1 : 1
    latitude = latitude + 2^(latitude_data(u));
end;
latitude

%*** longitude
longitude_data = coordinate_data(65:96);
longitude = 0;
for u = 32 : -1 : 1
    longitude = longitude + 2^(longitude_data(u));
end;
longitude

%*** speed
speed_data = coordinate_data(97:112);
speed = 0;
for u = 16 : -1 : 1
    speed = speed + 2^(speed_data(u));
end;
speed

%*** course
course_data = coordinate_data(113:128);
course = 0;
for u = 16 : -1 : 1
    course = course + 2^(course_data(u));
end;
course

File: AIS/ais.cpp
#include "stdafx.h"
#include <iostream>
#include <fstream>
#include <iomanip>
#include <stdlib.h>
#include <complex>
#include <math.h>
#include <cmath>
#include <algorithm>
#include "fasttransforms.h"

#include <string>
#include "ais.h"


// ФУНКЦИЯ ОСУЩЕСТВЛЯЕТ ОТКРЫТИЕ ФАЙЛА И ЧТЕНИЕ ИЗ НЕГО
void open_read_file(float data[], char file_name[])
{
    std::ifstream f;
    f.open(file_name);

    if(f)
    {
        int i = 0;
        while(!f.eof())
        {
            f >> data[i];
            i++;
        }
        f.close();
    }
    else
    {
        std::cout << "Error openning file" << std::endl;
    }
}
//*******************************************************************
// ПОЛУЧЕНИЕ КОМПЛЕСНОГО СИГНАЛА
void complex_sig(float data[], std::complex <float> *com, int size_signal)
{
    int l = 0;
    for( int j = 0; j < size_signal; j = j + 2)
    {
        com[l] = std::complex <float> (data[j], data[j+1]);
        l++;
    }
}
//*******************************************************************
// ПРИВЕДЕНИЕ СИГНАЛА К ДИФФЕРЕНЦИАЛЬНОМУ ВИДУ
void diff_complex_sig(std::complex <float> *com, std::complex <float> *diff_com, int size_signal)
{

    for(int k = 0; k < size_signal/2 ; k++)
    {
        diff_com[k] = com[k]*conj(com[k+1]);
    }
}
//*******************************************************************
// ДЕЛЕНИЕ КОМПЛЕКСНОЙ ПЛОСКОСТИ
void twd(std::complex <float> *twiddles)
{
    std::complex <float> one(0, 1);
    for(int t = 1; t < 18; t++)
    {
         twiddles[t] = exp(one * float((t - 1) *(M_PI/8)));
    }
}
//*******************************************************************
// КОРРЕЛЯЦИОННОЕ ОБНАРУЖЕНИЕ ПАКЕТА
void correlation(std::complex <float> synchro[], std::complex <float> *com,
                 alglib::complex *pIFFT, int length_frame, int size_signal)
{
    //ifft(fft(diff_com).*conj(fft(synchro)));
    //******************* РАЗБИВАЕМ НА ЭТАПЫ ************************

    //******************* FFT(SYNCHRO) ******************************
    alglib::complex_1d_array z;
    alglib::complex *pSynchro = new alglib::complex[length_frame];
    for(int u = 0; u < 128; u++){
        pSynchro[u].x = synchro[u].real();
        pSynchro[u].y = synchro[u].imag();
    }
    //*******************ЗАПОЛНЕНИЕ ПРЕАМБУЛЫ НУЛЯМИ ****************
    for(int u = 128; u < length_frame; u++){
        pSynchro[u].x = 0;
        pSynchro[u].y = 0;
    }
    z.setcontent(1024, pSynchro);
    fftc1d(z);
    //printf("%s ",z.tostring(3).c_str()); // СОВПАДАЕТ С MATLAB
    pSynchro = z.getcontent();

    //******************* CONJ(FFT(SYNCHRO)) ************************
    for(int o = 0; o < length_frame; o++)
    {
        pSynchro[o] = alglib::conj(pSynchro[o]);
    }

    //******** IFFT(FFT(DIFF_COM).*CONJ(FFT(SYNCHRO)))***************
    alglib::complex_1d_array y;
    alglib::complex_1d_array e;
    alglib::complex *pSignal = new alglib::complex[length_frame];
    alglib::complex *pMult = new alglib::complex[length_frame];
    alglib::complex *p = new alglib::complex[length_frame];

    std::ofstream fout;
    fout.open("ifft.txt");
    int r = 0;
    for(r = 0; r < size_signal/2-1; r = r+length_frame)
    {
        for(int u = 0; u < length_frame; u++){
        pSignal[u].x = com[u+r].real();
        pSignal[u].y = com[u+r].imag();
        }

        //************ FFT(DIFF_COM) ********************************
        y.setcontent(length_frame, pSignal);
        fftc1d(y);
        pSignal = y.getcontent();

        //********** FFT(DIFF_COM).*CONJ(FFT(SYNCHRO)) **************
        for( int h = 0; h <  length_frame; h++)
        {
            pMult[h].x = pSynchro[h].x*pSignal[h].x - pSynchro[h].y*pSignal[h].y;
            pMult[h].y = pSynchro[h].y*pSignal[h].x + pSynchro[h].x*pSignal[h].y;
        }

        //*********** IFFT() ****************************************
        e.setcontent(length_frame, pMult);
        fftc1dinv(e);
        p = e.getcontent();

        //*********** ЗАПИСЫВАЕМ В БОЛЬШОЙ ВЕКТОР *******************
        int k = 0;
        for(int i = r; i < r+length_frame; i++)
        {
            pIFFT[i] = p[k];
            fout << "i = " << i+1 << " : " << pIFFT[i].tostring(3).c_str() << std::endl;
            k++;
        }
    }
    fout.close();
}
//*******************************************************************
// ПОИСК МАКСИМАЛЬНОГО ЗНАЧЕНИЯ КОРРЕЛЯЦИИ В КАЖДОМ ФРЕЙМЕ
void find_max(alglib::complex *pIFFT, alglib::complex *max, int index[], int size_signal, int length_frame)
{
    int u = 0;
    for(int r = 0; r < size_signal/2-1; r = r+length_frame)
    {
        index[u] = 0;
        for(int i = r; i < r+length_frame; i++)
        {
            if(alglib::abscomplex(pIFFT[i]) > alglib::abscomplex(max[u]))
            {
                max[u] = pIFFT[i];
                index[u] = i;
            }
        }
        u++;
    }
}
//*******************************************************************
// ПОИСК МАКСИМАЛЬНОГО ЗНАЧЕНИЯ КОРРЕЛЯЦИИ СРЕДИ ФРЕЙМОВ, НОРМИРОВАНИЕ ЗНАЧЕНИЙ КОРРЕЛЯЦИИ
void norm_max(alglib::complex *max, int index[], alglib::complex *norm_max_corr)
{
    alglib::complex Max_corr = (0, 0);
    int id = 0;
    for(int i = 0; i < 16; i++)
    {
        if(alglib::abscomplex(max[i]) > alglib::abscomplex(Max_corr))
        {
           Max_corr = max[i];
        }
    }
    for(int i = 0; i < 16; i++)
    {
        norm_max_corr[i] = max[i]/Max_corr;
    }
}
//*******************************************************************
// ИЗВЛЕЧЕНИЕ ПАКЕТА ИЗ СИГНАЛА
void pars_package(alglib::complex *norm_max_corr, int index[], std::complex <float> *com,
                  std::complex <float> *package, int length_data)
{
    int number_index = 0;
    alglib::complex m(0,0);
    for(int i = 0; i < 16; i++)
    {
        if(alglib::abscomplex( norm_max_corr[i]) > alglib::abscomplex(m))
        {
           m =  norm_max_corr[i];
           number_index = index[i];
        }
    }
    for(int i = 0; i < length_data; i++)
    {
        package[i] = com[number_index+i];
    }
}
//*******************************************************************
// ДЕМОДУЛЯЦИЯ ПАКЕТА
void demodulation(std::complex <float> *package, int bits[], int length_data)
{
    std::complex <float> one(0, 1);
    std::complex <float> *sig1 = new std::complex <float> [5];
    std::complex <float> *sig2 = new std::complex <float> [5];
    sig1[0] = 1;
    sig1[1] = exp(one*float(M_PI/8));
    sig1[2] = exp(one*float(M_PI/4));
    sig1[3] = exp(one*float(3*M_PI/8));
    sig1[4] = one;
    for(int i = 0; i < 5; i++)
    {
        sig2[i] = std::conj(sig1[i]);
    }
    std::complex <float> *x = new std::complex <float> [5];
    int ii = 0;
    for(int i = 2; i < length_data; i = i + 4)
    {
        std::complex <float> pr1(0,0);
        std::complex <float> pr2(0,0);

        x[0] = package[i];
        x[1] = package[i+1];
        x[2] = package[i+2];
        x[3] = package[i+3];
        x[4] = package[i+4];
        for(int step = 0; step < 5; step++)
        {
            pr1 = pr1 + x[step]*sig1[step];
            pr2 = pr2 + x[step]*sig2[step];
        }

        if(pr1.real() > pr2.real())
        {
            bits[ii] = 0;
        }
        else
        {
            bits[ii] = 1;
        }
        ii++;
    }
    std::cout << "preamble ";
    for( int i = 0; i < 32; i++)
    {
        if(bits[i] == bits[i+1])
            bits[i] = 1;
        else
            bits[i] = 0;
        std::cout << bits[i] << " ";
    }
    std::cout << std::endl;
}
//*******************************************************************
// ПЕРЕВОД СЕГМЕНТА ДАННЫХ ПАКЕТА АИС ИЗ 2ой СИСТЕМЫ СЧИСЛЕНИЯ В 10ую
void translation(int coordinate_data[])
{
    int id_data[32];
    int latitude[32];
    int longitude[32];
    for(int i = 0; i < 32; i++)
    {
        id_data[i] = coordinate_data[i];
        latitude[i] = coordinate_data[32+i];
        longitude[i] = coordinate_data[64+i];
    }

    int ID = 0;
    int lat = 0;
    int lon = 0;
    for(int i = 31; i >= 0; i--)
    {
        ID = ID + pow(2,id_data[i]);
        lat = lat + pow(2,latitude[i]);
        lon = lon + pow(2,longitude[i]);
    }
    std::cout << "ID = " << ID << ", latitude = " << lat << ", longitude = " << lon;
    std::ofstream of("test.txt");
    if (!of) {
        std::cout << "Cannot open file.\n";
    }
    of << ID << " " << lat << " " << lon << " ";

    int speed_data[16];
    int course_data[16];
    for(int i = 0; i < 16; i++)
    {
        speed_data[i] = coordinate_data[96+i];
        course_data[i] = coordinate_data[112+i];
    }

    int speed = 0;
    int course = 0;
    for(int i = 15; i >= 0; i--)
    {
        speed = speed + pow(2,speed_data[i]);
        course = course + pow(2, course_data[i]);
    }
    std::cout << " speed = " << speed << ", course = " << course << std::endl;
    of << speed << " " << course;
    of.close();
}
//*******************************************************************
//****************** МОДУЛЬ АИС *************************************
// ВЫПОЛНЯЕТ ПОСЛЕДОВАТЕЛЬНУЮ ПОЛНУЮ ОБРАБОТКУ ПАКЕТА
void module_ais(int size_signal, int length_frame, int length_data, char file_name[])
{
    //*********** ОТКРЫТИЕ ФАЙЛА ************************************
    float data[size_signal];
    open_read_file(data, file_name);

    //**************** СКЛЕИВАНИЕ В КОМПЛЕКСНЫЕ ЧИСЛА ***************
    std::complex <float> *com = new std::complex  <float> [size_signal/2];
    complex_sig(data, com, size_signal);

    //***************** ПРИВЕДЕНИЕ К ДИФФ. ВИДУ *********************
    std::complex <float> *diff_com = new std::complex  <float> [size_signal/2];
    diff_complex_sig(com, diff_com, size_signal);

    //****************** ПОДГОТОВКА ПРЕАМБУЛЫ ***********************
    std::complex <float> *twiddles = new std::complex <float> [18];
    twd(twiddles);
    std::complex <float> synchro[] = {twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[4], twiddles[3], twiddles[2], twiddles[1], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[6], twiddles[7], twiddles[8], twiddles[9], twiddles[10], twiddles[11], twiddles[12], twiddles[13],
                                 twiddles[14], twiddles[15], twiddles[16], twiddles[17], twiddles[2], twiddles[3], twiddles[4], twiddles[5],
                                 twiddles[6], twiddles[7], twiddles[8], twiddles[9], twiddles[8], twiddles[7], twiddles[6], twiddles[5]};


    //****************** БЫСТРАЯ КОРЕЕЛЯЦИЯ *************************
    alglib::complex *pIFFT = new alglib::complex[size_signal/2];
    correlation(synchro, com, pIFFT, length_frame, size_signal);

    //*********************** ПОИСК МАКСИМУМА В КАЖДОМ ФРЕЙМЕ *******
    alglib::complex *max = new alglib::complex[16];
    int index[32];
    find_max(pIFFT, max, index, size_signal, length_frame);

    //********************** НОРМИРОВАНИЕ ***************************
    //******************** ПОИСК САМОГО МАКСИМАЛЬНОГО ЭЛЕМЕНТА ******
    alglib::complex *norm_max_corr = new alglib::complex[16];
    norm_max(max, index, norm_max_corr);

    //************ ВЫДЕЛЕНИЕ ПАКЕТА *********************************
    std::complex <float> *package = new std::complex  <float> [length_data];
    pars_package(norm_max_corr, index, com, package, length_frame);

    //********************** ДЕМОДУЛЯТОР ****************************
    int bits[length_data/4];
    demodulation(package, bits, length_data);

    //******************** ПАРСИНГ **********************************
    int coordinate_data[168];
    for(int i = 0; i < 168; i++)
    {
        coordinate_data[i] = bits[32+i];
    }

    //****************** ПЕРЕВОД В 10ую СИСТЕМУ *********************
    translation(coordinate_data);

}
